### ISSUE TITLE
Add support for complex interface

### ISSUE THREAD

**Issue by blegat** [jump-dev/SeDuMi.jl#23]
See https://github.com/jump-dev/JuMP.jl/pull/3257#issuecomment-1448418410 for a good MWE to use. I won't have time to take a look this week though

**Comment by araujoms** [comment]
I would be interested in adding proper complex support through MOI. I would be very helpful if you could give me some pointers on what needs to be done.

**Comment by araujoms** [comment]
Thanks a lot for the detailed pointers. I'm afraid I didn't understand everything. Isn't 1) just allowing complex numbers in the equality constraints? But what is the benefit of doing that directly in SeDuMi? What SeDuMi will do is just split the constraint into real and imaginary parts, which is what JuMP is already doing.

As for 2), isn't the ScaledPSDCone in MOI defined only for real symmetric matrices? Also, why would I want the version scaled by sqrt(2), since SeDuMi wants the non-scaled version as input?

**Comment by araujoms** [comment]
Ok, I see, `SeDuMi.ScaledPSDCone` is just how you defined the input for SeDuMi specifically, it has nothing to do with the MOI cones. That's why you wanted a bridge from the MOI cone (which is what JuMP will produce) to it.

SeDuMi does definitely want the matrix in square form as input. I also know that internally it will discard the upper triangular and re-scale the lower triangular. I would assume that doing that on the input would wreak havoc, but I guess it will become obvious why it works as I start tinkering. Thanks again!

**Comment by blegat** [comment]
Yes, we want to do everything as bridge so that it works well with MatrixOfConstraints and we can just take the matrix as is

**Comment by blegat** [comment]
`convert(Vector{MathOptInterface.VectorAffineTerm{ComplexF64}}, x)` ?

**Comment by araujoms** [comment]
That works, thanks. I thought that was equivalent to `ComplexF64.(x)`, which doesn't work, so I didn't even try.

**Comment by blegat** [comment]
Could you show me the code and the syntax error ? Maybe it's easier if you open a PR with your current code and show the error in the comment of the PR. I added you write access so that you can create a branch on the upstream remote directly so that I can modify it as well

**Comment by araujoms** [comment]
Well the syntax errors are caused by your code alone, you can literally paste that into `MOI_wrapper.jl` and see them. Some I could fix, but not all.

**Comment by araujoms** [comment]
Thanks for giving me write access, but accepting it requires me to enable two-factor authentication, which I refuse. I can do a regular PR instead if you want.

**Comment by odow** [comment]
> Thanks for giving me write access, but accepting it requires me to enable two-factor authentication, which I refuse

Any reason? This is a standard security policy for all repositories in jump-dev. Note that you don't have to use a phone number as the 2FA. See https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication

**Comment by araujoms** [comment]
Because 2FA is a huge pain in the ass, I'd rather keep using just the password to login.

**Comment by odow** [comment]
> Because 2FA is a huge pain in the ass, I'd rather keep using just the password to login

From our perspective, 2FA is to minimize the risk of someone's account being compromised and used to push malicious code to one of the repo's in JuMP-dev.

But also, the 2FA is pretty minimal. I stay logged into my account on my computer, so the only time I ever need to validate is performing an admin permission, like moving a repository, or changing secrets. You don't need to validate every time you open a PR or commit, so it shouldn't impact your day-to-day workflow.

**Comment by araujoms** [comment]
It is out of the question, I'm not going to mess up my github account, specially not in order to contribute to somebody else's project.

Moreover, the idea that someone would be interested in pushing malicious code to jump-dev repositories strikes me as paranoid. Such a commit would be instantly reverted and never become part of a release. And if you think my account is insecure you're welcome to try and hack it.

**Comment by blegat** [comment]
> I added you write access so that you can create a branch on the upstream remote directly so that I can modify it as well

It's fine if you push to your fork and create a PR, forget about it ;)

> Some I could fix, but not all.

Since you could fix some, I'd rather start from your branch than from my untested code

**Comment by araujoms** [comment]
I've already submitted a PR. And I did check the box "Allow edits by maintainers", so you have write access to it.

**Comment by blegat** [comment]
Yes, I missed that thanks, I'm going to check next week

**Comment by blegat** [comment]
Just pushed to your branch, it seems to go through, you should have what you need in `Ac_complex` now

**Comment by araujoms** [comment]
Awesome, thanks, indeed the data now gets to where it needs to be.
